### PR TITLE
Add `client_factory` parameter to `auth` methods

### PR DESCRIPTION
**Pull request by jlumbroso**
This is a pull request to implement a documented feature described in https://github.com/burnash/gspread/issues/1074

I have reviewed the submitting guidelines and I believe this follows all guidelines. (The exception is that it does not provide coverage: However unless I am mistaken there seems to be no tests for this functionality yet, and while this is a noble project, it's not the focus of this pull-request 😁 This feature is passive, it introduces no functionality, it is just a conduit to use the existing `gspread.client.BackoffClient`). Happy to make any modification necessary!

**Is your feature request related to a problem? Please describe.**
As noted, Google introduced [usage limits](https://developers.google.com/sheets/api/limits). While the behavior seems to not be well defined everywhere (in Drive API, the quota limit return 403 apparently undistinguishable by error code of an auth problem, for instance), in Sheets API the rate limit issue consistently returns 429. The existing `gspread.client.BackoffClient` class deals with this problem elegantly and effectively, and at the cost for the developer of one line change. Unfortunately the `auth` methods use `gspread.client.Client()` internally, in a hard-coded fashion.

**Describe the solution you'd like**
The feature introduces a parameter that is retrocompatible, as it defaults to the current behavior, but it also enables use of the `gspread.client.BackoffClient`.

**Describe alternatives you've considered**
- Modifying my local copy of the package only
- Catching the error at the higher level of my own code, but that seems sub-optimal, as it requires much more error catching code

**Additional context**
This is a small tweak, that is fully documented, that allows for better use of existing functionality of this package.

### CODE REVIEW

**Review by lavigne958 — COMMENTED** [review]
Hi think you for this PR. I've been thinking about adding this client for a while. Your approach is nice because backward compatible.

Give me some time to think if this new client is production ready then I'll merge this PR 🙃

**Comment by jlumbroso** [comment]
@lavigne958 Hey, sorry about the linter warnings, I just fixed them with the last commit, so your toolchain doesn't tell us the sky is falling! 😁

I think determining whether the client is production-ready will be a difficult decision, because it is not really: If it were production ready, then it should be the default, because that is clearly what Google wants (for all clients to behave parsimoniously to not have expensive peaks of requests).

On the other hand, right now, it is not **possible** to use the client at all, without hacking the library locally (which is what I had to do, to get around this issue in one of my projects, see https://github.com/josephlou5/roseingrave/issues/9). So my feature doesn't push the `BackoffClient` to the front, just opens a back door for power users to road test it — precisely so you can determine when it is production ready.

Taking this decision right now in the abstract is too difficult, and I would not be surprised if it's impossible to resolve!

**Comment by lavigne958** [comment]
You're right we should open the door for advanced users to use the `BackOffClient` and see how it goes.

I planed for release 6.0.0 to introduce it as defaults client so that would be a good way to start before that release.

**Review by lavigne958 — APPROVED** [review]

**Comment by jlumbroso** [comment]
Glad I could sway you 😊

Should be ready to merge, though let me know if you want any other changes.

**Comment by lavigne958** [comment]
For the good of everyone I don't monde changing my mind 😄.

I don't want much but the CI seems against you. You still have issues. You should run the CI locally first before pushing.

It's a simple `tox` command.

Once it's all green, we can merge it.

**Comment by jlumbroso** [comment]
For future reference, here is how I ran a local environment:
```
$ pipenv install tox tox-asdf
$ for v in 3.7.13 3.8.13 3.9.13 3.10.5; do asdf install python $v; done
$ pipenv run tox
$ pipenv run tox -e build
$ pipenv run tox -e docs
```
I would have gone crazy without:
https://github.com/apihackers/tox-asdf

Also, the final problem I was hunting down presented as this error:
```
Warning, treated as error:
/Users/jlumbroso/Programming/gspread/gspread/auth.py:docstring of gspread.auth.oauth:57:Field list ends without a blank line; unexpected unindent.
```

What it turned out to be is this (INCORRECT) docstring:
```
    :param :class:`gspread.ClientFactory` client_factory: A factory function that
        returns a client class. Defaults to :class:`gspread.Client` (but could also use
        :class:`gspread.BackoffClient` to avoid rate limiting)
```
into:
```
    :type client_factory: :class:`gspread.ClientFactory`
    :param client_factory: A factory function that returns a client class.
        Defaults to :class:`gspread.Client` (but could also use
        :class:`gspread.BackoffClient` to avoid rate limiting)
```
although the linter is tolerant to the form `:param <simple type> <param name>:` syntax, apparently for more complex return types it just does not work.

I also added a few fixes, like an export of `BackoffClient` in the top-level package, to mirror status of `Client`. And I also extended/uniformized the `gspread.authorize()` method.

I think everything should be good now, though I am crossing my fingers. 😁 🤞🏻

File: gspread/__init__.py
Comment (lavigne958) [inline gspread/__init__.py:38]: This should be in `Client` and not `client_factory` and in fact the previous comment was completely wrong. 
The full sphinix docstring is `` :class:`gspread.client.Client` ``

File: gspread/auth.py
Comment (lavigne958) [inline gspread/auth.py:171]: same here it should be: `` :class:`gspread.client.Client` ``

File: gspread/auth.py
Comment (lavigne958) [reply gspread/auth.py:251]: and same here it should be `` :class:`gspread.client.Client` ``

File: gspread/auth.py
Comment (lavigne958) [inline gspread/auth.py:291]: here too, the last part of the line should be: `` :class:`gspread.client.Client` ``

File: gspread/auth.py
Comment (lavigne958) [inline gspread/auth.py:296]: here too (basicaly ever occurences of the type ClientFactory in docstring should be updated with `` :class:`gspread.client.Client` ``

File: gspread/auth.py
Comment (lavigne958) [inline gspread/auth.py:325]: and this is the last one :slightly_smiling_face: to update with `` :class:`gspread.client.Client` ``

**Review by lavigne958 — COMMENTED** [review]
I noticed the changes you made about the return type `oauth_from_dict`, it returns a tuple and you removed it and replaced it with a single return type (the client type). :open_mouth: 

This method returns a tuple and should still mention it in the doc.

Then I noticed the return type you use and I checked out you branch and generated the doc, it works with no warnings but it does not resolve the link to the client class, if you use the docstring I mentioned on each occurrences then when generating the doc, each time it is mentioned you can click on the return type to reach the documentation about that type :wow: :sunglasses: 

Sorry I did not notice before, and this should be updated please :slightly_smiling_face:

File: gspread/__init__.py
Comment (jlumbroso) [reply gspread/__init__.py:42]: No problem.

File: gspread/__init__.py
Comment (jlumbroso) [reply gspread/__init__.py:38]: - I think `authorize()` is just top-level syntactic sugar over the functionality of `Client`. It needs to remain in the top-level. I can move it to `auth.py` if you want.

- Yes, I fixed the docstring to have the `:rtype:` and to clarify the explanatory string in the `:returns:`.

File: gspread/auth.py
Comment (jlumbroso) [reply gspread/auth.py]: Changed, sorry for introducing this bug with copy-pasting.

File: gspread/__init__.py
Comment (jlumbroso) [reply gspread/__init__.py]: I moved `authorize()` to `auth.py`

File: gspread/auth.py
Comment (jlumbroso) [reply gspread/auth.py]: Done.

File: gspread/auth.py
Comment (jlumbroso) [reply gspread/auth.py:305]: Here this is correct, and should be `ClientFactory` (which is `Type[Client]`), see:
https://docs.python.org/3/library/typing.html#typing.Type

Comment (jlumbroso) [reply gspread/auth.py:305]: Let me know if you still want this changed to `Client`.

File: gspread/auth.py
Comment (jlumbroso) [reply gspread/auth.py]: Changed.

File: gspread/auth.py
Comment (jlumbroso) [reply gspread/auth.py]: Changed!! 😁

**Comment by lavigne958** [comment]
> For future reference, here is how I ran a local environment:
> 
> ```
> $ pipenv install tox tox-asdf
> $ for v in 3.7.13 3.8.13 3.9.13 3.10.5; do asdf install python $v; done
> $ pipenv run tox
> $ pipenv run tox -e build
> $ pipenv run tox -e docs
> ```

I don't understand why you should go through such process, you can simply do `tox -e decs` and it should work.
The CI runs it using many different version of Python just case we miss something and to have a greater covering but I never experienced any issue with running the tests/linter/docs using only my local version of python (python-3.10.5).

I keep in my this too `tox-asdf` that can be useful, thank you for this tips.

> 
> I would have gone crazy without: https://github.com/apihackers/tox-asdf
> 
> Also, the final problem I was hunting down presented as this error:
> 
> ```
> Warning, treated as error:
> /Users/jlumbroso/Programming/gspread/gspread/auth.py:docstring of gspread.auth.oauth:57:Field list ends without a blank line; unexpected unindent.
> ```
> 
> What it turned out to be is this (INCORRECT) docstring:
> 
> ```
>     :param :class:`gspread.ClientFactory` client_factory: A factory function that
>         returns a client class. Defaults to :class:`gspread.Client` (but could also use
>         :class:`gspread.BackoffClient` to avoid rate limiting)
> ```
> 
> into:
> 
> ```
>     :type client_factory: :class:`gspread.ClientFactory`
>     :param client_factory: A factory function that returns a client class.
>         Defaults to :class:`gspread.Client` (but could also use
>         :class:`gspread.BackoffClient` to avoid rate limiting)
> ```
> 
> although the linter is tolerant to the form `:param <simple type> <param name>:` syntax, apparently for more complex return types it just does not work.

That's true, I never thought about it, never even noticed but their is not docstring with a complex type in the for you mention above. nice catch !

> 
> I also added a few fixes, like an export of `BackoffClient` in the top-level package, to mirror status of `Client`. And I also extended/uniformized the `gspread.authorize()` method.

that's nice, simpler for the users.

> 
> I think everything should be good now, though I am crossing my fingers. grin 🤞🏻

yeah.... still not green :laughing:  but we'll get there, I promise you have my unconditional attention now, we need to merge this PR. My new priority for this project.

**Comment by jlumbroso** [comment]
@lavigne958 Thanks for your patience with this pull-request. The linter had an issue because I moved `authorize` from `__init__.py` to `auth.py`, but then that meant the import statement in `__init__.py` was too long and had to be changed from:
```python
from .auth import authorize, oauth, oauth_from_dict, service_account, service_account_from_dict
```
to:
```python
from .auth import (
    authorize,
    oauth,
    oauth_from_dict,
    service_account,
    service_account_from_dict,
)
```
I reran everything locally, hoping to prevent not green check, but let me know 😊

**Review by lavigne958 — APPROVED** [review]
LGTM ! :+1:

**Comment by lavigne958** [comment]
Well done @jlumbroso  and thank you for your patience and your contribution.

**Comment by lavigne958** [comment]
closes #1074

**Comment by jlumbroso** [comment]
Thank you @lavigne958 ! So glad I could help! Thanks for taking an interest in this feature, and for helping me get the PR up to standards! 🙏🏻😊

**Pull request merged**
